python/xorbits/_mars/dataframe/sort/psrs.py

```diff
@@ -660,15 +660,77 @@ def _execute_series_map(cls, ctx, op):
                 values = a.iloc[poses[i] : poses[i + 1]]
                 ctx[out.key, (i,)] = values
 
+    @classmethod
+    def _calc_pivots_poses(
+        cls, index, pivots, is_multi_index=False, level=None, sort_remaining=True
+    ):
+        if not is_multi_index:
+            return index.searchsorted(list(pivots), side="right")
+
+        if index.nlevels != pivots.nlevels:
+            raise TypeError(
+                f"Number of levels should be the same between index and pivots. But got nlevels of index is {index.nlevels}, and nlevels of pivots is {pivots.nlevels}"
+            )
+
+        if isinstance(level, list) and len(level) == 1 and not sort_remaining:
+            return index.get_level_values(level[0]).searchsorted(
+                list(pivots.get_level_values(level[0])), side="right"
+            )
+        else:
+            if level is None:
+                index_list = list(index)
+                pivots_list = list(pivots)
+            elif isinstance(level, list):
+                index_from_level = []
+                pivots_from_level = []
+                for l in level:
+                    index_from_level.append(index.get_level_values(l))
+                    pivots_from_level.append(pivots.get_level_values(l))
+                if sort_remaining:
+                    if len(level) < index.nlevels:
+                        remain_indexes = index.droplevel(level)
+                        remain_pivots = pivots.droplevel(level)
+                        for l in range(remain_indexes.nlevels):
+                            index_from_level.append(remain_indexes.get_level_values(l))
+                        for l in range(remain_pivots.nlevels):
+                            pivots_from_level.append(remain_pivots.get_level_values(l))
+                index_list = list(zip(*index_from_level))
+                pivots_list = list(zip(*pivots_from_level))
+            else:
+                raise TypeError(
+                    f"level should be list or None after processing, but got type {type(level)}."
+                )
+
+            poses = []
+            for pivot in pivots_list:
+                poses.append(bisect.bisect_right(index_list, pivot))
+
+            return np.array(poses)
+
     @classmethod
     def _execute_sort_index_map(cls, ctx, op):
         a, pivots = [ctx[c.key] for c in op.inputs]
         out = op.outputs[0]
 
+        if isinstance(a.index, pd.MultiIndex) and isinstance(pivots, pd.MultiIndex):
+            is_multi_index = True
+        elif not isinstance(a.index, pd.MultiIndex) and not isinstance(
+            pivots, pd.MultiIndex
+        ):
+            is_multi_index = False
+        else:
+            raise TypeError(
+                f"It can only handle the case where a.index and pivots are both MultiIndex or neither are MultiIndex. But get type of a.index is {type(a.index)} and type of pivots is {type(pivots)}"
```

Comment:
Better make error message more consise. Let ChatGPT do it for you :)

Reply:
done, thx